### ISSUE TITLE
CUDA_EXP  takes double as much GPU memory as GPU mode.

### ISSUE THREAD

**Issue by bluesummerv** [microsoft/LightGBM#5318]
## Description
CUDA_EXP  takes double as much GPU memory as GPU mode.

## Environment info
linux    python3.9
LightGBM version or commit hash:
latest   ff947bfac5148a9b94d045f2879aee183e02e174

Thanks！  @shiyu1994 

```shell

```

**Comment by jameslamb** [comment]
Thanks very much for your interest in LightGBM!

Are you able to provide any more details, like a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) showing how you determined this? It's possible that only some part of the CUDA_EXP implementation has a different memory usage profile than GPU mode (for example, maybe you are using categorical features with high cardinality or something).

Without some sample code, it'll be very difficult for maintainers here to investigate this claim.

**Comment by bluesummerv** [comment]
`
n1=4000000
n2 = 2200
x = np.random.random((n1,n2))
y = np.random.random((n1))
dataset = lgb.Dataset(data=x,lable=y,free_raw_data=True)
dataset.construct()
dataset.save_binary(path)
dataset = lgb.Dataset(path)

params ={}
params['device']  = 'gpu'
params['num_leaves']  = 100
params['num_iterations']  = 100
params['learning_rate']  = 0.02
params['metric']  = 'mse'

m = lgb.train(params, dataset)
`
